python/tutorials/11-programmatic-dependent-launch.py

```diff
@@ -0,0 +1,265 @@
+        # If we utilize Programmatic Dependent Launch, we must wait on the prior kernel
+        # to complete in case a or b are written to by the prior kernel.
+        # This is done to prevent races.
+        tl.extra.cuda.gdc_wait()
+
+    if USE_TMA:
+        # We have no guarentees about what memory may be passed for our make tensor descriptor,
+        # and whether there is inter kernel dependencies. We therefore must safely allocate after
+        # Waiting on the prior kernel to finish.
+        a_desc = tl.make_tensor_descriptor(
+            a_ptr,
+            shape=[M, K],
+            strides=[K, 1],
+            block_shape=[BLOCK_SIZE_M, BLOCK_SIZE_K],
+        )
+        b_desc = tl.make_tensor_descriptor(
+            b_ptr,
+            shape=[N, K],
+            strides=[K, 1],
+            block_shape=[BLOCK_SIZE_N, BLOCK_SIZE_K],
+        )
+        c_desc = tl.make_tensor_descriptor(
+            c_ptr,
+            shape=[M, N],
+            strides=[N, 1],
+            block_shape=[BLOCK_SIZE_M, BLOCK_SIZE_N],
+        )
+
+    for ki in range(k_tiles):
+        a = None
+        b = None
+        if USE_TMA:
+            offs_k = ki * BLOCK_SIZE_K
+            a = a_desc.load([offs_am, offs_k])
+            b = b_desc.load([offs_bn, offs_k])
+        else:
+            a = tl.load(a_ptrs, mask=offs_k[None, :] < K - ki * BLOCK_SIZE_K, other=0.0)
+            b = tl.load(b_ptrs, mask=offs_k[None, :] < K - ki * BLOCK_SIZE_K, other=0.0)
+            a_ptrs += BLOCK_SIZE_K
+            b_ptrs += BLOCK_SIZE_K
+        accumulator = tl.dot(a, b.T, accumulator)
+
+    if USE_GDC:
+        # GDC launch dependents is used to launch dependent kernels.
+        # Once GDC launch it is possible for the next kernel to begin if
+        # there are enough resources.
+        tl.extra.cuda.gdc_launch_dependents()
+
+    c = accumulator.to(dtype)
+    if USE_TMA:
+        c_desc.store([offs_am, offs_bn], c)
+    else:
+        offs_cm = pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+        offs_cn = pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
+        c_ptrs = c_ptr + N * offs_cm[:, None] + offs_cn[None, :]
+        c_mask = (offs_cm[:, None] < M) & (offs_cn[None, :] < N)
+        tl.store(c_ptrs, c, mask=c_mask)
+
+
+def matmul(a, b, launch_pdl=True):
+    # Check constraints.
+    assert a.shape[1] == b.shape[1], "Incompatible dimensions"  # b is transposed
+    assert a.dtype == b.dtype, "Incompatible dtypes"
+
+    M, K = a.shape
+    N, K = b.shape
+    dtype = a.dtype
+
+    c = torch.empty((M, N), device=a.device, dtype=dtype)
+
+    # TMA descriptors require a global memory allocation
+    def alloc_fn(size: int, alignment: int, stream: Optional[int]):
+        return torch.empty(size, device="cuda", dtype=torch.int8)
+
+    triton.set_allocator(alloc_fn)
+
+    grid = lambda META: (triton.cdiv(M, META["BLOCK_SIZE_M"]) * triton.cdiv(N, META["BLOCK_SIZE_N"]), )
+    matmul_kernel[grid](
```

Comment:
How do the benchmarks change if we use a persistent kernel? It seems kind of pointless to say we can get a huge speedup when you use a really inefficient kernel to begin with.

Reply:
Hypothetically, it may still be a good speedup since the work partitioning is entirely static for a persistent matmul so there may be more idle SMs while one kernel ramps down. Would be interesting to see.

Reply:
The intent here was to focus on a single wave of computation. So, persistence does not matter in that case. I can add a persistent version as well.

Reply:
Oh I see, the speedup comes from launching such a small size that it doesn't fill the GPU SMs. This is less interesting IMO. Would you not see an even bigger speedup from using multiple streams instead?

Reply:
The issue with multiple streams is if there are dependencies between kernels. For example if GEMM 1 feeds into GEMM 2, both GEMMs cannot run in parallel. PDL allows for these to overlap for parts of the kernel that are independent (e.g., prologue of kernel 2 does not depend on kernel one). For small back to back GEMMs (memory bandwidth bound) that are only a few waves, this is a very useful optimization to have.